### PR TITLE
TF.exe

### PR DESCRIPTION
**Pull request by ericsciple**

### CODE REVIEW

**Comment by ericsciple** [comment]
Not ready for merge.

File: src/Microsoft.VisualStudio.Services.Agent/JobServer.cs [hunk 1]
```diff
@@ -38,6 +38,20 @@ public sealed class JobServer : AgentService, IJobServer
             if (!_connection.HasAuthenticated)
             {
                 await _connection.ConnectAsync();
+                // TODO: REMOVE THIS TRACE STATEMENT BEFORE COMMITING TO MASTER.
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/JobServer.cs]: remove

File: src/Agent.Worker/Build/TFCommandManager.cs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+            }
+        }
+
+        protected override string Switch => "/";
+
+        protected override string TF
+        {
+            get
+            {
+                return Path.Combine(
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/TFCommandManager.cs]: Don't forget this TODO :D

File: src/Agent.Worker/Build/TFCommandManager.cs [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        {
+            get
+            {
+                return TfsVCFeatures.DefaultWorkfoldMap |
+                    TfsVCFeatures.LoginType |
+                    TfsVCFeatures.Scorch;
+            }
+        }
+
+        protected override string Switch => "/";
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/TFCommandManager.cs]: ArgumentsSeparator ?

File: src/Agent.Worker/Build/TfsVCSourceProvider.cs [hunk 4]
```diff
@@ -75,36 +73,37 @@ public sealed class TfsVCTeeSourceProvider : SourceProvider, ISourceProvider
...
 
-            // Attempt to re-use an existing workspace if clean=false.
-            TeeWorkspace existingTeeWorkspace = null;
+            // Attempt to re-use an existing workspace if the command manager supports scorch
+            // or if clean is not specified.
+            ITfsVCWorkspace existingTFWorkspace = null;
             bool clean = endpoint.Data.ContainsKey(WellKnownEndpointData.Clean) &&
                 StringUtil.ConvertToBoolean(endpoint.Data[WellKnownEndpointData.Clean], defaultValue: false);
-            if (!clean)
+            if (tf.Features.HasFlag(TfsVCFeatures.Scorch) || !clean)
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/TfsVCSourceProvider.cs]: the logic around clean is a little bit weird.

File: src/Agent.Worker/Build/TfsVCSourceProvider.cs [hunk 5]
```diff
@@ -259,26 +259,24 @@ public override string GetLocalPath(IExecutionContext executionContext, ServiceE
...
+                (definitionMapping.LocalPath ?? string.Empty)
+                .Trim('/', '\\');
+            if (Path.DirectorySeparatorChar == '\\')
+            {
+                relativePath = relativePath.Replace('/', Path.DirectorySeparatorChar);
+            }
+            else
+            {
+                relativePath = relativePath.Replace('\\', Path.DirectorySeparatorChar);
+            }
```
Comment (TingluoHuang) [inline src/Agent.Worker/Build/TfsVCSourceProvider.cs]: why we need replace '\' and '/' for linux?

**Pull request merged**